pkg/storage/driver/sql.go

```diff
@@ -0,0 +1,330 @@
+		      		created_at INTEGER NOT NULL,
+		      		modified_at INTEGER NOT NULL DEFAULT 0
+            );
+
+            CREATE INDEX ON releases (key);
+            CREATE INDEX ON releases (version);
+            CREATE INDEX ON releases (status);
+            CREATE INDEX ON releases (owner);
+            CREATE INDEX ON releases (created_at);
+            CREATE INDEX ON releases (modified_at);
+		      `,
+				},
+				Down: []string{
+					`
+            DROP TABLE releases;
+          `,
+				},
+			},
+		},
+	}
+
+	_, err := migrate.Exec(s.db.DB, "postgres", migrations, migrate.Up)
+	return err
+}
+
+// Release describes a Helm release
+type Release struct {
+	Key  string `db:"key"`
+	Body string `db:"body"`
+
+	Name       string `db:"name"`
+	Version    int    `db:"version"`
+	Status     string `db:"status"`
+	Owner      string `db:"owner"`
+	CreatedAt  int    `db:"created_at"`
+	ModifiedAt int    `db:"modified_at"`
+}
+
+// NewSQL initializes a new memory driver.
+func NewSQL(dialect, connectionString string, logger func(string, ...interface{})) (*SQL, error) {
+	if _, ok := supportedSQLDialects[dialect]; !ok {
+		return nil, fmt.Errorf("%s dialect isn't supported, only \"postgres\" is available for now", dialect)
+	}
+
+	db, err := sqlx.Connect(dialect, connectionString)
+	if err != nil {
+		return nil, err
+	}
+
+	driver := &SQL{
+		db:  db,
+		Log: logger,
+	}
+
+	if err := driver.ensureDBSetup(); err != nil {
+		return nil, err
+	}
+
+	return driver, nil
+}
+
+// Get returns the release named by key.
+func (s *SQL) Get(key string) (*rspb.Release, error) {
+	var record Release
+	// Get will return an error if the result is empty
+	err := s.db.Get(&record, "SELECT body FROM releases WHERE key = $1", key)
+	if err != nil {
+		s.Log("got SQL error when getting release %s: %v", key, err)
+		return nil, storageerrors.ErrReleaseNotFound(key)
+	}
+
+	release, err := decodeRelease(record.Body)
+	if err != nil {
+		s.Log("get: failed to decode data %q: %v", key, err)
+		return nil, err
+	}
+
+	return release, nil
+}
+
+// List returns the list of all releases such that filter(release) == true
+func (s *SQL) List(filter func(*rspb.Release) bool) ([]*rspb.Release, error) {
+	var records = []Release{}
+	if err := s.db.Select(&records, "SELECT body FROM releases WHERE owner = 'TILLER'"); err != nil {
+		s.Log("list: failed to list: %v", err)
+		return nil, err
+	}
+
+	var releases []*rspb.Release
+	for _, record := range records {
+		release, err := decodeRelease(record.Body)
+		if err != nil {
+			s.Log("list: failed to decode release: %v: %v", record, err)
+			continue
+		}
+		if filter(release) {
+			releases = append(releases, release)
+		}
+	}
+
+	return releases, nil
+}
+
+// Query returns the set of releases that match the provided set of labels.
+func (s *SQL) Query(labels map[string]string) ([]*rspb.Release, error) {
+	var sqlFilterKeys []string
+	sqlFilter := map[string]interface{}{}
+	for key, val := range labels {
```

Comment:
You might need to sanitize the value of `key` to avoid a Bobby Tables attack. https://www.xkcd.com/327/

Reply:
We actually check that `key` is in this (sane) map before executing the generated SQL request: https://github.com/helm/helm/pull/5371/files#diff-9614ccdf050917eda8876d24c7cb63ceR37 at line 197. Would that be fine to you ?

Reply:
yeah, that looks fine to me!